src/event/inspect.rs

```diff
@@ -0,0 +1,74 @@
+use crate::event::{Event, LogEvent};
+use parking_lot::RwLock;
+use std::collections::HashMap;
+use tokio::sync::broadcast;
+
+type Result = ::std::result::Result<Event, ()>;
+type Sender = broadcast::Sender<LogEvent>;
+type Receiver = broadcast::Receiver<LogEvent>;
+
+pub struct EventInspector {
+    components: RwLock<HashMap<String, Sender>>,
+}
+
+impl EventInspector {
+    pub fn new() -> Self {
+        Self {
+            components: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Associates a broadcast::Sender with a component name, so that subscribers may selectively
+    /// receive `LogEvent`s based on specific components. Returns a function that can be provided
+    /// to `.inspect` on an upstream LogEvent receiver channel.
+    pub fn adder(&self, component_name: String) -> impl FnMut(&Event) {
+        let mut lock = self.components.write();
+        let tx = lock.entry(component_name).or_insert_with(|| {
+            let (tx, _) = broadcast::channel(100);
+            tx
+        });
+
+        // Clone the sender, to move an owned copy into the returning function
+        let tx = tx.clone();
+
+        move |ev| {
+            if let Event::Log(ev) = ev {
+                // broadcast::Sender provides an atomic count of the number of active
+                // listeners. We're using that here to avoid a potentally expensive clone
+                // in the (likely) event that there are no current subscribers.
+                //
+                // This can suffer from TOC/TOU, but the risk is minimal as the purpose
+                // here is solely to reduce expense.
+                if tx.receiver_count() > 0 {
```

Comment:
@MOZGIII, this is the guard that's done to check for active listeners. I'm making the (very presumptive, completely unbenchmarked) assumption that a `std::sync::Mutex` lock around the listeners is cheaper than cloning the LogEvent (and all of its serde Value's) and using `try_send` or equivalent.

There are undoubtedly cheaper ways to do this. Doing our own local checks against the HashMap that stores component names might be one of them. I reached for the out-the-box option tokio provides for simplicity, but totally open to improving this.

Reply:
cc benchmarking king @jszwedko -- you might have some ideas here too.